core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -611,24 +872,56 @@ class ReplicaManager(val config: KafkaConfig,
     }
   }
 
-  def localLogOrException(topicPartition: TopicPartition): Log = {
-    getPartitionOrException(topicPartition).localLogOrException
+  // Online + Fenced
+  def getNonOfflinePartitionOrError(topicPartition: TopicPartition): Either[Errors, Partition] = {
+    getPartition(topicPartition) match {
+      case HostedPartition.Online(partition) =>
+        Right(partition)
+
+      case HostedPartition.Fenced(partition) =>
+        Right(partition)
+
+      case HostedPartition.Offline =>
+        Left(Errors.KAFKA_STORAGE_ERROR)
+
+      case HostedPartition.None if metadataCache.contains(topicPartition) =>
+        // The topic exists, but this broker is no longer a replica of it, so we return NOT_LEADER_OR_FOLLOWER which
+        // forces clients to refresh metadata to find the new location. This can happen, for example,
+        // during a partition reassignment if a produce request from the client is sent to a broker after
+        // the local replica has been deleted.
+        Left(Errors.NOT_LEADER_OR_FOLLOWER)
+
+      case HostedPartition.None =>
+        Left(Errors.UNKNOWN_TOPIC_OR_PARTITION)
+    }
+  }
+
+  def localOnlineLogOrException(topicPartition: TopicPartition): Log = {
```

Comment:
Could we please add a jmh benchmark for the frequently called ReplicaManager log lookup methods? Many of these are called very frequently in hot paths such as the fetch API code or replica fetcher, and we need to make sure we don't regress as the code is already overly heavy - I had an experimental patch to cut down on the use of Either and it dropped the CPU time and garbage by half. Unfortunately I can't locate the experiment or I would point you to the benchmark. I'm fine if we don't change the approach here as long as we have the benchmark and we aren't regressing.